components/builder-api/src/server/framework/middleware.rs

```diff
@@ -98,12 +99,62 @@ fn authenticate(req: &HttpRequest<AppState>, token: &str) -> error::Result<Sessi
 
     // Check for a valid personal access token
     if bldr_core::access_token::is_access_token(token) {
-        let session =
-            bldr_core::access_token::validate_access_token(&req.state().config.api.key_path, token)
-                .map_err(|_| NetError::new(ErrCode::BAD_TOKEN, "net:auth:bad-token"))?;
-        revocation_check(req, session.get_id(), token)
-            .map_err(|_| NetError::new(ErrCode::BAD_TOKEN, "net:auth:revoked-token"))?;
-        return Ok(session);
+        let mut memcache = req.state().memcache.borrow_mut();
+        match memcache.get_session(token) {
+            Some(session) => {
+                trace!("Session {} Cache Hit!", token);
+                return Ok(session);
+            }
+            None => {
+                trace!("Session {} Cache Miss!", token);
+                // Pull the session out of the current token provided so we can validate
+                // it against the db's tokens
+                let mut session = bldr_core::access_token::validate_access_token(
+                    &req.state().config.api.key_path,
+                    token,
+                ).map_err(|_| NetError::new(ErrCode::BAD_TOKEN, "net:auth:bad-token"))?;
+                // If we can't find a token in the cache, we need to round-trip to the
+                // db to see if we have a valid session token. The first time this is used,
+                // it is only to get the account_id so we can do the lookup
+                match do_get_access_tokens(&req, session.get_id()) {
+                    Ok(access_tokens) => match access_tokens.get_tokens().first() {
```

Comment:
assert access_tokens length == 1

Reply:
What does that net us if we are matching on some or none for `.first()`

Reply:
It's not a huge deal - but it would be a false negative if there were multiple tokens and the second token was the matching one.

Reply:
well it may be even better to change do_get_access_tokens to do_get_access_token that only returns one .. we probably put in more infra for multiple tokens than we need right now